libs/libc/machine/Kconfig

```diff
@@ -137,6 +137,54 @@ config LIBC_ARCH_ELF_64BIT
 	default n
 	depends on LIBC_ARCH_ELF
 
+config LIBC_ARCH_PREVENT_KERNEL
+	bool
+	default n
+	---help---
+		Prevent the architecture-specific libc from being used in the kernel.
+
+config LIBC_ARCH_PREVENT_USER
+	bool
+	default n
+	---help---
+		Prevent the architecture-specific libc from being used in the
+		userspace.
+
+if BUILD_PROTECTED || BUILD_KERNEL
+
+choice LIBC_ARCH_USAGE
```

Comment:
What's the point of exposing this as a user-selectable option?
Isn't the motivation for the proposed change to overcome a limitation of some chips? If so, this will only increase the maintenance burden.

Reply:
Choices can't be hidden. If we try to:

```
warning: <choice LIBC_ARCH_USAGE> (defined at libs/libc/machine/Kconfig:155) defined without a prompt
```

Furthermore, I think it's desirable to provide users a way to select whether to use arch's libc only in the kernel/userspace  (if not "prevented" by the chip) depending on their applications.

Reply:
Sorry, that's not what I meant. Instead I mean removing the choice completely.
```c
#if (defined(CONFIG_LIBC_ARCH_PREVENT_KERNEL) && defined(__KERNEL__)) || \
    (defined(CONFIG_LIBC_ARCH_PREVENT_USER) && !defined(__KERNEL__))
```

> Furthermore, I think it's desirable to provide users a way to select whether to use arch's libc only in the kernel/userspace (if not "prevented" by the chip) depending on their applications.

I disagree. Protected mode has been there for ages and nobody had this necessity, so most probably it will never be used.
Again, this will only increase the maintenance burden.

Reply:
> Sorry, that's not what I meant. Instead I mean removing the `choice` completely and use `LIBC_ARCH_PREVENT_[KERNEL/USER]` directly.
> 
> ```c
> #if (defined(CONFIG_LIBC_ARCH_PREVENT_KERNEL) && defined(__KERNEL__)) || \
>     (defined(CONFIG_LIBC_ARCH_PREVENT_USER) && !defined(__KERNEL__))
> ```
> 
> > Furthermore, I think it's desirable to provide users a way to select whether to use arch's libc only in the kernel/userspace (if not "prevented" by the chip) depending on their applications.
> 
> I disagree. Protected mode has been there for ages and nobody had this necessity, so most probably it will never be used. Again, this will only increase the maintenance burden.

Well, Espressif "has" this necessity. I chose to select `CONFIG_LIBC_ARCH_PREVENT_USER` at chip's config, but strictly speaking, we only need this behavior when the Wi-Fi driver is being used. Thus, alternatively,  `CONFIG_LIBC_ARCH_PREVENT_USER` could be only selected by the Wi-Fi driver or, otherwise, the choice be defined by a `defconfig` (and the user could tweak this choice accordingly).

Particularly, I disagree with the maintenance burden: I think it brings flexibility and I prefer to keep it. 

@xiaoxiang781216 suggested a similar approach. What do you think about it?

Reply:
> Well, Espressif "has" this necessity.
> we only need this behavior when the Wi-Fi driver is being used.

This is not accurate.
The Wi-Fi driver is the only one that **so far** requires it because of the way it uses the libc functions. BTW, I faced the same issue with `memcpy` when implementing Protected Mode support for ESP32-S3.
https://github.com/apache/nuttx/blob/ea276b20ef9edede2c90a4e7fd708b45b4bc01d4/arch/xtensa/src/esp32s3/esp32s3_userspace.c#L548-L559

So it is desirable to avoid creating pitfalls for the unaware user.

Reply:
Done!